From 2ea8b5d236286f043f52781318afaf5a6ed3b9c6 Mon Sep 17 00:00:00 2001
From: Nirbheek Chauhan <nirbheek@centricular.com>
Date: Mon, 10 Nov 2014 22:26:09 +0530
Subject: [PATCH 2/6] owr/owr.c: Register gstnice as an ordinary gst plugin

---
 owr/owr.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/owr/owr.c b/owr/owr.c
index 8b5977a..169976c 100644
--- a/owr/owr.c
+++ b/owr/owr.c
@@ -48,8 +48,6 @@
 
 #ifdef OWR_STATIC
 #include <stdlib.h>
-
-#include <gstnice.h>
 #endif
 
 #ifdef __APPLE__
@@ -77,6 +75,7 @@ GST_PLUGIN_STATIC_DECLARE(coreelements);
 GST_PLUGIN_STATIC_DECLARE(erdtls);
 GST_PLUGIN_STATIC_DECLARE(inter);
 GST_PLUGIN_STATIC_DECLARE(mulaw);
+GST_PLUGIN_STATIC_DECLARE(nice);
 GST_PLUGIN_STATIC_DECLARE(opengl);
 GST_PLUGIN_STATIC_DECLARE(openh264);
 GST_PLUGIN_STATIC_DECLARE(opus);
@@ -204,6 +203,7 @@ void owr_init()
     GST_PLUGIN_STATIC_REGISTER(erdtls);
     GST_PLUGIN_STATIC_REGISTER(inter);
     GST_PLUGIN_STATIC_REGISTER(mulaw);
+    GST_PLUGIN_STATIC_REGISTER(nice);
     GST_PLUGIN_STATIC_REGISTER(opengl);
     GST_PLUGIN_STATIC_REGISTER(openh264);
     GST_PLUGIN_STATIC_REGISTER(opus);
@@ -237,8 +237,6 @@ void owr_init()
     GST_PLUGIN_STATIC_REGISTER(ercolorspace);
 #endif
 
-    gst_element_register(NULL, "nicesrc", GST_RANK_NONE, GST_TYPE_NICE_SRC);
-    gst_element_register(NULL, "nicesink", GST_RANK_NONE, GST_TYPE_NICE_SINK);
 #endif
 
     owr_main_context_is_external = !!owr_main_context;
-- 
2.1.0

