From 7e4022665964b5485dc3a03e15092185d63b7629 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian@centricular.com>
Date: Wed, 3 Dec 2014 19:32:39 +0100
Subject: [PATCH 3/6] owr: There is no g_io_module_load_static()

---
 libopenwebrtc.exp | 1 -
 1 file changed, 1 deletion(-)

diff --git a/libopenwebrtc.exp b/libopenwebrtc.exp
index e0f9bbf..b1b800d 100644
--- a/libopenwebrtc.exp
+++ b/libopenwebrtc.exp
@@ -1610,7 +1610,6 @@ g_io_extension_point_register
 g_io_extension_point_set_required_type
 g_io_extension_ref_class
 g_io_module_get_type
-g_io_module_load_static_module
 g_io_module_new
 g_io_module_scope_block
 g_io_module_scope_flags_get_type
-- 
2.1.0

