From 0e2bb357ff75053240b380e0d8657e25072f764b Mon Sep 17 00:00:00 2001
From: Nirbheek Chauhan <nirbheek@centricular.com>
Date: Tue, 17 Feb 2015 05:13:08 +0530
Subject: [PATCH 5/6] libopenwebrtc.exp: Remove unused non-public symbols

---
 libopenwebrtc.exp | 78 +------------------------------------------------------
 1 file changed, 1 insertion(+), 77 deletions(-)

diff --git a/libopenwebrtc.exp b/libopenwebrtc.exp
index b1b800d..8109e26 100644
--- a/libopenwebrtc.exp
+++ b/libopenwebrtc.exp
@@ -150,14 +150,6 @@ g_application_get_is_remote
 g_application_get_type
 g_application_hold
 g_application_id_is_valid
-g_application_impl_activate
-g_application_impl_command_line
-g_application_impl_destroy
-g_application_impl_flush
-g_application_impl_get_dbus_connection
-g_application_impl_get_dbus_object_path
-g_application_impl_open
-g_application_impl_register
 g_application_new
 g_application_open
 g_application_quit
@@ -472,7 +464,6 @@ g_charset_converter_get_use_fallback
 g_charset_converter_new
 g_charset_converter_set_use_fallback
 g_chdir
-g_check_setuid
 g_checksum_copy
 g_checksum_free
 g_checksum_get_digest
@@ -721,7 +712,6 @@ g_date_valid_weekday
 g_date_valid_year
 g_dbus_action_group_get
 g_dbus_action_group_get_type
-g_dbus_action_group_sync
 g_dbus_address_escape_value
 g_dbus_address_get_for_bus_sync
 g_dbus_address_get_stream
@@ -1007,11 +997,6 @@ g_dbus_signal_info_ref
 g_dbus_signal_info_unref
 g_dbus_subtree_flags_get_type
 g_dcgettext
-g_delayed_settings_backend_apply
-g_delayed_settings_backend_get_has_unapplied
-g_delayed_settings_backend_get_type
-g_delayed_settings_backend_new
-g_delayed_settings_backend_revert
 g_desktop_app_info_get_boolean
 g_desktop_app_info_get_categories
 g_desktop_app_info_get_filename
@@ -1409,7 +1394,6 @@ g_get_user_data_dir
 g_get_user_name
 g_get_user_runtime_dir
 g_get_user_special_dir
-g_get_worker_context
 g_getenv
 g_gstring_get_type
 g_gtype_get_type
@@ -1729,9 +1713,7 @@ g_local_file_monitor_get_type
 g_locale_from_utf8
 g_locale_to_utf8
 g_log
-g_log_always_fatal
 g_log_default_handler
-g_log_msg_prefix
 g_log_remove_handler
 g_log_set_always_fatal
 g_log_set_default_handler
@@ -1755,7 +1737,6 @@ g_main_context_invoke_full
 g_main_context_is_owner
 g_main_context_iteration
 g_main_context_new
-g_main_context_new_with_next_id
 g_main_context_pending
 g_main_context_pop_thread_default
 g_main_context_prepare
@@ -1853,7 +1834,6 @@ g_memory_output_stream_new
 g_memory_output_stream_new_resizable
 g_memory_output_stream_steal_as_bytes
 g_memory_output_stream_steal_data
-g_memory_settings_backend_get_type
 g_memory_settings_backend_new
 g_menu_append
 g_menu_append_item
@@ -2028,7 +2008,6 @@ g_node_push_allocator
 g_node_reverse_children
 g_node_traverse
 g_node_unlink
-g_null_settings_backend_get_type
 g_null_settings_backend_new
 g_nullify_pointer
 g_object_add_toggle_ref
@@ -2399,7 +2378,6 @@ g_resolver_error_quark
 g_resolver_free_addresses
 g_resolver_free_targets
 g_resolver_get_default
-g_resolver_get_serial
 g_resolver_get_type
 g_resolver_lookup_by_address
 g_resolver_lookup_by_address_async
@@ -2518,25 +2496,13 @@ g_setenv
 g_settings_apply
 g_settings_backend_changed
 g_settings_backend_changed_tree
-g_settings_backend_create_tree
 g_settings_backend_flatten_tree
 g_settings_backend_get_default
-g_settings_backend_get_permission
 g_settings_backend_get_type
-g_settings_backend_get_writable
 g_settings_backend_keys_changed
 g_settings_backend_path_changed
 g_settings_backend_path_writable_changed
-g_settings_backend_read
-g_settings_backend_reset
-g_settings_backend_subscribe
-g_settings_backend_sync_default
-g_settings_backend_unsubscribe
-g_settings_backend_unwatch
-g_settings_backend_watch
 g_settings_backend_writable_changed
-g_settings_backend_write
-g_settings_backend_write_tree
 g_settings_bind
 g_settings_bind_flags_get_type
 g_settings_bind_with_mapping
@@ -2552,7 +2518,6 @@ g_settings_get_flags
 g_settings_get_has_unapplied
 g_settings_get_int
 g_settings_get_mapped
-g_settings_get_mapping
 g_settings_get_range
 g_settings_get_string
 g_settings_get_strv
@@ -2564,7 +2529,6 @@ g_settings_list_children
 g_settings_list_keys
 g_settings_list_relocatable_schemas
 g_settings_list_schemas
-g_settings_mapping_is_compatible
 g_settings_new
 g_settings_new_full
 g_settings_new_with_backend
@@ -2573,24 +2537,12 @@ g_settings_new_with_path
 g_settings_range_check
 g_settings_reset
 g_settings_revert
-g_settings_schema_get_gettext_domain
 g_settings_schema_get_id
+g_settings_schema_get_key
 g_settings_schema_get_path
-g_settings_schema_get_string
 g_settings_schema_get_type
-g_settings_schema_get_value
 g_settings_schema_has_key
-g_settings_schema_key_clear
-g_settings_schema_key_from_enum
-g_settings_schema_key_from_flags
-g_settings_schema_key_get_translated_default
-g_settings_schema_key_init
 g_settings_schema_key_range_check
-g_settings_schema_key_range_fixup
-g_settings_schema_key_to_enum
-g_settings_schema_key_to_flags
-g_settings_schema_key_type_check
-g_settings_schema_list
 g_settings_schema_ref
 g_settings_schema_source_get_default
 g_settings_schema_source_get_type
@@ -2605,7 +2557,6 @@ g_settings_set_double
 g_settings_set_enum
 g_settings_set_flags
 g_settings_set_int
-g_settings_set_mapping
 g_settings_set_string
 g_settings_set_strv
 g_settings_set_uint
@@ -2740,7 +2691,6 @@ g_slist_reverse
 g_slist_sort
 g_slist_sort_with_data
 g_snprintf
-g_socket
 g_socket_accept
 g_socket_address_enumerator_get_type
 g_socket_address_enumerator_next
@@ -3043,11 +2993,6 @@ g_strtod
 g_strup
 g_strv_get_type
 g_strv_length
-g_system_thread_exit
-g_system_thread_free
-g_system_thread_new
-g_system_thread_set_name
-g_system_thread_wait
 g_task_attach_source
 g_task_get_cancellable
 g_task_get_check_cancellable
@@ -3157,7 +3102,6 @@ g_thread_init_glib
 g_thread_init_with_errorcheck_mutexes
 g_thread_join
 g_thread_new
-g_thread_new_internal
 g_thread_pool_free
 g_thread_pool_get_max_idle_time
 g_thread_pool_get_max_threads
@@ -3172,7 +3116,6 @@ g_thread_pool_set_max_unused_threads
 g_thread_pool_set_sort_function
 g_thread_pool_stop_unused_threads
 g_thread_pool_unprocessed
-g_thread_proxy
 g_thread_ref
 g_thread_self
 g_thread_set_priority
@@ -3695,7 +3638,6 @@ g_variant_get_size
 g_variant_get_string
 g_variant_get_strv
 g_variant_get_type
-g_variant_get_type_info
 g_variant_get_type_string
 g_variant_get_uint16
 g_variant_get_uint32
@@ -3709,7 +3651,6 @@ g_variant_is_normal_form
 g_variant_is_object_path
 g_variant_is_of_type
 g_variant_is_signature
-g_variant_is_trusted
 g_variant_iter_copy
 g_variant_iter_free
 g_variant_iter_init
@@ -3731,7 +3672,6 @@ g_variant_new_dict_entry
 g_variant_new_double
 g_variant_new_fixed_array
 g_variant_new_from_bytes
-g_variant_new_from_children
 g_variant_new_from_data
 g_variant_new_handle
 g_variant_new_int16
@@ -3851,11 +3791,6 @@ g_volume_should_automount
 g_vprintf
 g_vsnprintf
 g_vsprintf
-g_wakeup_acknowledge
-g_wakeup_free
-g_wakeup_get_pollfd
-g_wakeup_new
-g_wakeup_signal
 g_warn_message
 g_weak_ref_clear
 g_weak_ref_get
@@ -3880,14 +3815,3 @@ glib_micro_version
 glib_minor_version
 glib_on_error_halt
 glib_pgettext
-gvdb_table_get_raw_value
-gvdb_table_get_table
-gvdb_table_get_value
-gvdb_table_has_value
-gvdb_table_is_valid
-gvdb_table_list
-gvdb_table_new
-gvdb_table_new_from_data
-gvdb_table_ref
-gvdb_table_unref
-gvdb_table_walk
-- 
2.1.0

