From 1d0f97ec4cc92e0903d6c46b3434dc737780bef8 Mon Sep 17 00:00:00 2001
From: Nirbheek Chauhan <nirbheek@centricular.com>
Date: Mon, 10 Nov 2014 18:38:09 +0530
Subject: [PATCH] modules/DynamicObject: Add missing LIBS

---
 modules/DynamicObject/Makefile.am | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/modules/DynamicObject/Makefile.am b/modules/DynamicObject/Makefile.am
index 56536cb..c43262e 100644
--- a/modules/DynamicObject/Makefile.am
+++ b/modules/DynamicObject/Makefile.am
@@ -19,6 +19,9 @@ libseed_DynamicObject_la_LDFLAGS = \
 	$(GOBJECT_INTROSPECTION_LDFLAGS) \
 	$(SEED_PROFILE_LIBS)
 
+libseed_DynamicObject_la_LIBADD = \
+	$(top_builddir)/libseed/libseed@SEED_GTK_VERSION@.la
+
 endif
 
 
-- 
2.1.0

